### PR TITLE
frd:u: Initial stub some functions

### PR DESCRIPTION
**Pull request by mailwl**
It's minimal set to start the game Fantasy Life, most functions just stubbed (need more investigation)
according #901 this functions were implemented, but i don't see it in master

### CODE REVIEW

File: src/core/hle/service/frd/frd.cpp [hunk 1]
```diff
@@ -10,6 +12,69 @@
...
+    LOG_WARNING(Service_FRD, "(STUBBED) called");
+}
+
+void GetFriendKeyList(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 unknown   = cmd_buff[1];
+    u32 frd_count = cmd_buff[2];
+    u32 buf_addr  = cmd_buff[65];
+    FriendKey* frd_key_list = (FriendKey*)Memory::GetPointer(buf_addr);
```
Comment (lioncash) [inline src/core/hle/service/frd/frd.cpp]: reinterpret_cast

File: src/core/hle/service/frd/frd.cpp [hunk 2]
```diff
@@ -10,6 +12,69 @@
...
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_FRD, "(STUBBED) called");
+}
+
+void GetMyScreenName(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = 0; // No error
+    // TODO: (mailwl) get the name from config
+    Common::UTF8ToUTF16("Citra").copy((char16_t*)&cmd_buff[2], 20);
```
Comment (lioncash) [inline src/core/hle/service/frd/frd.cpp]: reinterpret_cast

File: src/core/hle/service/frd/frd.h [hunk 3]
```diff
@@ -4,9 +4,75 @@
 
 #pragma once
 
+#include "core/hle/service/service.h"
```
Comment (lioncash) [inline src/core/hle/service/frd/frd.h]: This can be a forward declaration

File: src/core/hle/service/frd/frd.cpp [hunk 4]
```diff
@@ -10,6 +12,101 @@
...
+
+void GetMyPresence(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 magic_value = cmd_buff[64];
+    u32 my_presence_addr = cmd_buff[65];
+
+    ASSERT(magic_value == 0x4B0002);
+
+    MyPresence* my_presence = reinterpret_cast<MyPresence*>(Memory::GetPointer(my_presence_addr));
```
Comment (Subv) [inline src/core/hle/service/frd/frd.cpp]: Do not use GetPointer, use Memory::WriteBlock

File: src/core/hle/service/frd/frd.cpp [hunk 5]
```diff
@@ -10,6 +12,101 @@
...
+
+void GetFriendKeyList(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 unknown       = cmd_buff[1];
+    u32 frd_count     = cmd_buff[2];
+    u32 frd_key_addr  = cmd_buff[65];
+
+    FriendKey* frd_key_list = reinterpret_cast<FriendKey*>(Memory::GetPointer(frd_key_addr));
+    memset(frd_key_list, 0, frd_count * sizeof(FriendKey));
```
Comment (Subv) [inline src/core/hle/service/frd/frd.cpp]: same here

File: src/core/hle/service/frd/frd.cpp [hunk 6]
```diff
@@ -10,6 +12,101 @@
...
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 count = cmd_buff[1];
+    u32 count_s = cmd_buff[2];
+    u32 frd_key_addr = cmd_buff[3];
+    u32 profiles_addr = cmd_buff[65];
+
+    ASSERT(count_s == (count << 18 | 2));
+
+    Profile* profiles = reinterpret_cast<Profile*>(Memory::GetPointer(profiles_addr));
```
Comment (Subv) [inline src/core/hle/service/frd/frd.cpp]: I'm going to stop pointing these out, but they should be changed everywhere they appear.

**Comment by mailwl** [comment]
done. also change function Interface::SyncRequest() a little (remove unnecessary else{})

File: src/core/hle/service/frd/frd.cpp [hunk 7]
```diff
@@ -10,6 +12,98 @@
...
+
+void GetMyPresence(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 shifted_out_size = cmd_buff[64];
+    u32 my_presence_addr = cmd_buff[65];
+
+    ASSERT(shifted_out_size == ((sizeof(MyPresence) << 14) | 2));
+
+    Memory::WriteBlock(my_presence_addr, reinterpret_cast<const u8*>(&my_presence), sizeof(MyPresence));
```
Comment (Subv) [inline src/core/hle/service/frd/frd.cpp]: There is no need for the reinterpret_cast when casting to u8*

File: src/core/hle/service/frd/frd.cpp [hunk 8]
```diff
@@ -10,6 +12,98 @@
 namespace Service {
 namespace FRD {
 
+typedef u16(display_name)[11];
```
Comment (Subv) [inline src/core/hle/service/frd/frd.cpp]: What.

File: src/core/hle/service/frd/frd.h [hunk 9]
```diff
@@ -4,9 +4,98 @@
...
+    u32 friend_id;
+    u32 unknown;
+    u64 friend_code;
+};
+
+struct MyPresence {
+    u8 unknown[0x12C];
+};
+
+struct Profile
```
Comment (Subv) [inline src/core/hle/service/frd/frd.h]: The style of these structures is inconsistent

File: src/core/hle/service/frd/frd.cpp (hunk 7 above)
Comment (mailwl) [inline src/core/hle/service/frd/frd.cpp]: vs 2015 says: can not convert type "MyPresence*" to "const u8*"

**Comment by bunnei** [comment]
@mailwl Do you plan on addressing the above comments?

**Comment by mailwl** [comment]
Yes, but i still don't know, how to convert struct pointer to const u8* without casting
`Memory::WriteBlock(my_presence_addr, reinterpret_cast<const u8*>(&my_presence), sizeof(MyPresence));`

**Comment by bunnei** [comment]
Looks fine to me.

File: src/core/hle/service/service.cpp [hunk 10]
```diff
@@ -70,9 +70,8 @@ ResultVal<bool> Interface::SyncRequest() {
         // TODO(bunnei): Hack - ignore error
         cmd_buff[1] = 0;
         return MakeResult<bool>(false);
-    } else {
-        LOG_TRACE(Service, "%s", MakeFunctionString(itr->second.name, GetPortName().c_str(), cmd_buff).c_str());
     }
+    LOG_TRACE(Service, "%s", MakeFunctionString(itr->second.name, GetPortName().c_str(), cmd_buff).c_str());
```
Comment (yuriks) [inline src/core/hle/service/service.cpp]: Was this intentional? This will generate a double-log when an error happens with TRACE log level.

Comment (mailwl) [inline src/core/hle/service/service.cpp]: No, there is no double log. `if` block always ends with `return MakeResult<bool>(false);`
resharper/c++ says "Redundant `else` keyword"

**Pull request merged**